File: prac1_sorts/c++ src/countingSort.cpp
#include <iostream>
#include "Timer.hpp"
using namespace std;
#define max 1000

void counting_sort(int* &array, int n)
{

    int *count = new int[max]{0};
    for (int i = 0; i < n; i++)
        count[array[i]]++;

    for (int i = 0; i < max-1 ; i++)
        count[i + 1] = count[i] + count[i + 1];

    for (int i = max-1; i > 0; i--)
        count[i] = count[i-1];
    count[0] = 0;

    int *newA = new int[n]{0};
    
    for (int i = 0; i < n; i++)
    {
        newA[count[array[i]]] = array[i];
        count[array[i]]++;
    }
    
    delete [] array;
    delete [] count;
    array = newA;
}

int main()
{
    int numLengths;
    cin>>numLengths;

    for (int i = 0; i < numLengths; ++i)
    {
        float avg = 0;
        int numArrays = 10;
        int n;
        cin>>n;
        for(int j = 0; j < numArrays; ++j)
        {
            int* arr = new int [n]{0};
            for (int k = 0; k < n; ++k)
                cin>>arr[k];
            Timer timer;
            timer.start();
            counting_sort(arr,n);
            timer.stop();
            avg += timer.elapsedMilliseconds();
            delete [] arr;
        }
        avg /= numArrays;
        cout << n << " " << avg << endl;
    }
    

    return 0;
}

File: prac1_sorts/python src/countingSort.py
from timer import Timer

filer = open('arrays.txt', "r")
filew = open("countingSortTime.txt","w")
#lectura de lineas
lines = []
for line in filer:
    lines.append(line)
filer.close()
#arreglo ordenado
array = []
max = 1000

def counting_sort(arr,n):
    count = [0 for i in range(max+1)]

    for i in range(n):
        count[arr[i]] += 1
 
    for i in range(max):
        count[i+1] = count[i] + count[i+1]

    for i in range(max,0,-1):
        count[i] = count[i-1]
    count[0] = 0
    
    newA = [0 for i in range(n)]

    for i in range(n):
        newA[count[arr[i]]] = arr[i]
        count[arr[i]] += 1
    array = newA


if __name__ == "__main__":
    t=Timer()
    index = 0
    numArrays = int(lines[index])
    for i in range(numArrays):
        arr = []
        index +=1
        n = int(lines[index])
        index +=1
        aux = lines[index].split(" ")
        for j in range(n):
            arr.append(int(aux[j]))
            
        t.start()
        counting_sort(arr,n)
        addtimer = t.stop()
        filew.writelines([str(n)," ",str(round(addtimer,5)),"\n"])  
    filew.close()